validator/client/src/main/java/tech/pegasys/teku/validator/client/DoppelgangerDetectionService.java

```diff
@@ -0,0 +1,181 @@
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.api.migrated.ValidatorLivenessAtEpoch;
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+import tech.pegasys.teku.infrastructure.async.Cancellable;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.time.TimeProvider;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.service.serviceutils.Service;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+import tech.pegasys.teku.validator.beaconnode.GenesisDataProvider;
+
+public class DoppelgangerDetectionService extends Service {
+
+  private static final Logger LOGGER = LogManager.getLogger();
+  private final Duration checkDelay;
+  private final Duration timeout;
+  private final AsyncRunner asyncRunner;
+  private final ValidatorApiChannel validatorApiChannel;
+  private final ValidatorIndexProvider validatorIndexProvider;
+  private final Spec spec;
+  private final TimeProvider timeProvider;
+  private final GenesisDataProvider genesisDataProvider;
+  private Optional<UInt64> epochAtStart = Optional.empty();
+  private volatile Cancellable doppelgangerDetectionTask;
+  private final AtomicBoolean doppelgangerCheckFinished;
+  private long startTime;
+
+  public DoppelgangerDetectionService(
+      final AsyncRunner asyncRunner,
+      final ValidatorApiChannel validatorApiChannel,
+      final ValidatorIndexProvider validatorIndexProvider,
+      final Spec spec,
+      final TimeProvider timeProvider,
+      final GenesisDataProvider genesisDataProvider,
+      final Duration checkDelay,
+      final Duration timeout) {
+    this.asyncRunner = asyncRunner;
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorIndexProvider = validatorIndexProvider;
+    this.spec = spec;
+    this.timeProvider = timeProvider;
+    this.genesisDataProvider = genesisDataProvider;
+    this.doppelgangerCheckFinished = new AtomicBoolean(false);
+    this.checkDelay = checkDelay;
+    this.timeout = timeout;
+  }
+
+  @Override
+  protected SafeFuture<?> doStart() {
+    startTime = System.nanoTime();
+    doppelgangerDetectionTask =
+        asyncRunner.runWithFixedDelay(
+            this::checkValidatorsDoppelganger,
+            checkDelay,
+            checkDelay,
+            throwable -> LOGGER.error("Error while checking validators doppelgangers.", throwable));
+    while (true) {
+      if (doppelgangerCheckFinished.get()) {
+        return SafeFuture.COMPLETE;
+      }
+    }
+  }
+
+  @Override
+  protected SafeFuture<?> doStop() {
+    LOGGER.info("No validator doppelganger detected. Stopping doppelganger detection service.");
```

Comment:
not sure we need an info here? doStop will be called any time the service is stopped so probably need to be selective about what the message says if we do have it...

eg. start then stop straight away will cause stop to be called and it'll report...